juju/juju
# LXDProfile - Modification Status

## Description of change

The follow adds a modification status, that allows us to display
when a change to a running machine fails, but we want the error
to be displayed to the operator without it disappearing.

So in the current implementation the status output uses last write
wins. Which has been good enough for most implementations up until
now, where by juju only catered for errors during provisioning.
The change here is a fundamental change, where we state that changes
during a machine update can affect what the operator sees in juju
status.

This was discussed with @jameinel for a possible solution. With
this solution we can now factor in new potential changes that happen
and can bubble them up to the operator with good effect.

The naming of the status "modification" isn't stuck in stone, but
it's to imply the outcome of an action to a machine. This can be
changed and open to suggestions.

I've had to bump the ProvisionerAPI to v8, so when looking at this
PR it's advised to look over that.

## QA steps

```sh
juju bootstrap lxd 
juju deploy ~/bundle-no-sub.yaml 
watch --color -n 1 juju status --color
```

Check that the error is kept when the machine `0` can't apply the
profile.

```
Model    Controller  Cloud/Region  Version  SLA          Timestamp
default  yyy         lxd/default   2.5.1.1  unsupported  17:30:12Z

App          Version  Status  Scale  Charm        Store  Rev  OS      Notes
lxd-profile  18.04    active      5  lxd-profile  local    0  ubuntu
ubuntu       18.04    active      5  ubuntu-lite  local    0  ubuntu

Unit            Workload  Agent  Machine  Public address  Ports  Message
lxd-profile/0*  active    idle   0        10.178.104.242         load: 5.50, 2.77, 3.26
lxd-profile/1   active    idle   1        10.178.104.171         load: 1.97, 2.22, 3.24
lxd-profile/2   active    idle   2        10.178.104.76          load: 1.97, 2.22, 3.24
lxd-profile/3   active    idle   3        10.178.104.55          load: 1.35, 1.97, 3.05
lxd-profile/4   active    idle   4        10.178.104.132         load: 1.35, 1.97, 3.05
ubuntu/0*       active    idle   0        10.178.104.242         load: 1.97, 2.22, 3.24
ubuntu/1        active    idle   1        10.178.104.171         load: 1.97, 2.22, 3.24
ubuntu/2        active    idle   2        10.178.104.76          load: 2.43, 2.31, 3.26
ubuntu/3        active    idle   3        10.178.104.55          load: 1.88, 2.21, 3.24
ubuntu/4        active    idle   4        10.178.104.132         load: 1.88, 2.21, 3.24

Machine  State    DNS             Inst id        Series  AZ  Message
0        error    10.178.104.242  juju-ad601b-0  bionic      cannot upgrade machine's lxd profile: Failed to add mount for device: Failed to run: /snap/lxd/current/bin/lxd forkmount lxc-mount juju-ad601b-0 /var/snap/lxd/common/lxd/containers /var/snap/lxd/common/lxd/logs/juju-ad601b-0/lxc.conf /var/snap/lxd/common/lxd/devices/juju-ad601b-0/unix.tun.dev-net-tun /dev/net/tun none 4096:
1        started  10.178.104.171  juju-ad601b-1  bionic      Running
2        started  10.178.104.76   juju-ad601b-2  bionic      Running
3        started  10.178.104.55   juju-ad601b-3  bionic      Running
4        started  10.178.104.132  juju-ad601b-4  bionic      Running
```

bundle-no-sub.yaml
```yaml
series: bionic
applications:
  lxd-profile:
    charm: ~/go/src/github.com/juju/juju/testcharms/charm-repo/quantal/lxd-profile
    num_units: 9
    to:
    - "0"
    - "1"
    - "2"
    - "3"
    - "4"
    - "5"
    - "6"
    - "7"
    - "8"
  ubuntu:
    charm: cs:~jameinel/ubuntu-lite
    num_units: 9
    to:
    - "0"
    - "1"
    - "2"
    - "3"
    - "4"
    - "5"
    - "6"
    - "7"
    - "8"
machines:
  "0": {}
  "1": {}
  "2": {}
  "3": {}
  "4": {}
  "5": {}
  "6": {}
  "7": {}
  "8": {}
```
